pkg/localstorev2/internal/chunkstore/txchunkstore_test.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package chunkstore_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/localstorev2/internal/chunkstore"
+	postagetesting "github.com/ethersphere/bee/pkg/postage/testing"
+	"github.com/ethersphere/bee/pkg/sharky"
+	chunktest "github.com/ethersphere/bee/pkg/storage/testing"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/storagev2/inmemstore"
+	"github.com/ethersphere/bee/pkg/storagev2/storagetest"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/spf13/afero"
+)
+
+func TestTxChunkStore(t *testing.T) {
+	t.Parallel()
+
+	sharky, err := sharky.New(&memFS{Fs: afero.NewMemMapFs()}, 1, swarm.SocMaxChunkSize)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	storagetest.TestTxChunkStore(t, chunkstore.NewTxChunkStore(inmemstore.NewTxStore(inmemstore.New()), sharky))
+}
+
+// TestMultipleStampsRefCnt tests the behaviour of ref counting along with multiple
+// stamps to ensure transactions work correctly.
+func TestMultipleStampsRefCnt(t *testing.T) {
+	t.Parallel()
+
+	sharky, err := sharky.New(&memFS{Fs: afero.NewMemMapFs()}, 1, swarm.SocMaxChunkSize)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	chunkStore := chunkstore.NewTxChunkStore(inmemstore.NewTxStore(inmemstore.New()), sharky)
+
+	chunk := chunktest.GenerateTestRandomChunk()
+	stamps := []swarm.Stamp{chunk.Stamp()}
+	for i := 0; i < 2; i++ {
+		stamps = append(stamps, postagetesting.MustNewStamp())
+	}
+
+	verify := func(t *testing.T) {
+		t.Helper()
+
+		rIdx := chunkstore.RetrievalIndexItem{
+			Address: chunk.Address(),
+		}
+
+		has, err := chunkStore.(*chunkstore.TxChunkStoreWrapper).Store().Has(&rIdx)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if !has {
+			t.Fatalf("retrievalIndex not found %s", chunk.Address())
+		}
+
+		for _, stamp := range stamps {
+			sIdx := chunkstore.ChunkStampItem{
+				Address: chunk.Address(),
+				Stamp:   stamp,
+			}
+
+			has, err := chunkStore.(*chunkstore.TxChunkStoreWrapper).Store().Has(&sIdx)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if !has {
+				t.Fatalf("chunkStampItem not found %s", chunk.Address())
+			}
+		}
+	}
+
+	t.Run("put with multiple stamps", func(t *testing.T) {
+		cs := chunkStore.NewTx(storage.NewTxState(context.TODO()))
+
+		for _, stamp := range stamps {
+			err := chunkStore.Put(context.TODO(), chunk.WithStamp(stamp))
+			if err != nil {
+				t.Fatalf("failed to put chunk: %v", err)
+			}
+		}
+
+		err := cs.Commit()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		verify(t)
+	})
+
+	t.Run("rollback delete operations", func(t *testing.T) {
```

Comment:
There should be tests for the rollback of the put operation when a new value is created and when an existing value is updated (probably as part fo the `storagetest.TestTxChunkStore` test suite).

Reply:
So in the chunkstore updates would generally not happen as these are content-addressed values. So if the key is the same the data will also be the same. The only thing updated could be the stamps.

The reason I have these tests here as they test the refCnt functionality which is specific to this chunkstore. But I guess we could add 1 more test in txchunkstore to Put with different stamps and then rollback keeps the right stamps.